### PR TITLE
dcv revision 2022

### LINKED ISSUE #129: is this library dead?
I have wanted to try my new [regionprops](https://github.com/aferust/regionpropsford) library with dcv. I could not even built dcv using dub + ldc2 (on both ubuntu and win10). It always complains about mir.algorithm no matter which version I use. Please say dcv is not dead. It would be awesome if we have something like opencv in dlang :(

### PR DESCRIPTION
**Pull request by aferust**
For potential reviews of @henrygouk , @ljubobratovicrelja, @9il.

### CODE REVIEW

**Comment by ljubobratovicrelja** [comment]
Hi @aferust, thanks so much for this.

A bit ashamed to say, I am downloading compilers :), and setting everything up to fetch this and try to understand better all of your changes. 

However, to help me in that, it would mean a lot if you would make like an outline of all your changes. Frankly, having that many changes in one commit is quite unwelcomed, and something that we would usually turn down for revision, as it is too complex and time consuming - having smaller and grouped changes in separate commits is highly preferred. However, in this instance, I am more than willing to put more effort into it, and understand everything. Just saying that for future references, I hope you don't mind.

Some concrete questions (please excuse if you may have mentioned those in our conversation in your issue post):

- Why is original project source organization changed - source/dcv content is extracted to the root of the project. Is that some update on dub side? (I am totally unfamiliar with what was happening in the last 2-3 years with D development)
- Why is ffmpeg injected like so? Why not prepare it like dub dependency? I know [the package that was previously used](https://github.com/ljubobratovicrelja/ffmpeg-d) does not support v4, and I can see that [new packages](https://code.dlang.org/packages/ffmpeg4d) have sprung in dub database, but I see this as a bit chaotic solution. We started organizing the ffmpeg-d repository to support different ffmpeg versions via separate branches (much like the original project), so I feel this should have been added there. However, I definitely am not lobbying for that, more like stating that this feels wrong. At the same time, I would think supporting 4.x versions within ffmpeg-d may end up being a lot of work, so in that case, I'd be totally fine with using the new dub package, but let's just not have them injected within this library like presented.
- I guess unrevised contains moved code that you mentioned earlier, that is yet to be adopted to new D/Mir? This is the stuff where you would expect some help from Henry's end?
- Why is video module extracted like so? Isn't is also supposed to be under 'unrevised'?

Note - as this is a work in progress (for e.g. because of the 'unrevised' content), I was thinking of proposing this to be merged in a specific branch instead of master. However, as master is not even buildable, and quite dead - let's continue and merge with master after revision. Then we can step by step try to bring this library back to life, and sort everything back, and make it neat. Let me know how does that sound.

Also, something I was working just prior to giving up on it, [this branch](https://github.com/libmir/dcv/tree/refactorzilla-pt1). Have you seen it maybe? It may be irrelevant, but I think it touches some of the aspects you've been thinking about regarding Slice vs Image structures. I don't know how relevant this is according to modern Mir API, but I guess the integration of [color](https://github.com/libmir/dcv/blob/refactorzilla-pt1/dub.json#L52) package, and how it is integrated within DCV (checkout [this](https://github.com/libmir/dcv/blob/refactorzilla-pt1/source/dcv/core/types.d) and also how [image IO is handled](https://github.com/libmir/dcv/blob/refactorzilla-pt1/source/dcv/io/image.d#L47-L152)), may be interesting to you, and I would hope that in some way we can end up using some of those changes. I personally very much liked the new API that relied on color package, because the work done there was very good, and by no means I would want to redo it in any way. I can see [that development there is also quite slow](https://github.com/TurkeyMan/color/commits/master), but I would think such library has to be nurtured, if thing like DCV should exist in D's ecosystem.

That's it, sorry for the long text. There's a lot of it left undone here, I guess. :) Again, thanks so much for taking interest in this project.

**Comment by aferust** [comment]
I am sorry for this PR. I know it is hard to follow with its present form. It is better you reject it, and we separate FFMpeg injection first. Your comments and questions made some unclear parts solid in my mind.

Things to do before the PR that can be reviewed easier.
- Removing FFmpeg injection (You and I will conclude a solution together, refer to my answer 2)
- Restructure source files to keep them in the source folder. This will also help review the modifications easily. I will handle this.

1>Why is the original project source organization changed - source/dcv content is extracted to the root of the project. Is that some update on dub side? (I am totally unfamiliar with what was happening in the last 2-3 years with D development)

- I did it by following the project structures such as dplug and arsd. I wanted this to enable separate imports, for instance, to import video and ffmpeg related things separately. However, I hadn't thought of that we can store the source files in a source folder.  I sometimes do such stupid things without considering the best option. I agree with you, we can use  "sourcePaths" of dub. I will make arrangements accordingly.

2> Why is ffmpeg injected like so? Why not prepare it like dub dependency?....
- This is my workaround. I totally agree with you.  "We started organizing the ffmpeg-d repository to support different ffmpeg versions via separate branches ".  I could not notice it. I now believe my modifications could be more visible if the FFmpeg binding was omitted. I don't want to add one more ffmpeg-d package to the dub registry by causing garbage of packages in dub.  I don't want to deeply dive into FFmpeg either. So, what can be done for a neat solution now? You have already stated that "supporting 4.x versions within ffmpeg-d may end up being a lot of work".  What about including dcv:ffmpeg441 to ffmpeg-d with a custom versioning such as "ffmpeg-d": "4.4.1+dcv" associating with an individual git branch? I believe that the dub registry allows this. But, if you are not into this, as you said, I can create a separate dub package for the FFmpeg441 binding.

3>I guess unrevised contains moved code that you mentioned earlier, that is yet to be adopted to new D/Mir? This is the stuff where you would expect some help from Henry's end?

- Yes. I do.  I actually touched that code a bit to comply with the new Slice! schema.

4> Why is video module extracted like so?

- I also wanted this to enable separate import. 

5>Isn't is also supposed to be under 'unrevised'?
- No, it is revised. It can read video frames from webcam stream and files. My webcam works with it. I don't have other cameras to test at the moment. Some video files are read with a distortion. I don't know why? maybe API changes in ffmpeg or pre-existing issue? Please check out the examples folder. All of them work with my modifications (except stereo). 

6> Note - as this is a work in progress (for e.g. because of the 'unrevised' content), I was thinking of proposing this to be merged in a specific branch instead of master. However, as master is not even buildable, and quite dead - let's continue and merge with master after revision. Then we can step by step try to bring this library back to life, and sort everything back, and make it neat. Let me know how does that sound.

Yes this approach is safer and comfy in many ways. Then, please let me know if I need to do something other than this PR. Should I make another PR to another branch (possible created by you). There are things that I have not done with git system so far like this.

7> Also, something I was working just prior to giving up on it, this branch. Have you seen it maybe? ...

- No, I have just seen it right now (my bad). I will examine it.

8> but I guess the integration of color package, and how it is integrated within DCV (checkout this and also how image IO is handled), may be interesting to you, and I would hope that in some way we can end up using some of those changes. I personally very much liked the new API that relied on color package,

- I will examine it too.

**Comment by ljubobratovicrelja** [comment]
Thanks for all your very elaborated answers!

If this isn't too big of a problem for you, I would advise we close this PR, and do the 2 points you state:
- Removing FFmpeg injection
- Restructure source files to keep them in the source folder. This will also help review the modifications easily. I will handle this.

I will help you with the first one, we can see how much work it would actually be to include it in ffmpeg-d. Maybe it isn't so scary, and it would be quite beneficial for ffmpeg-d to get a refresh.

As for source file organization, you mentioned that you restructured them like so to be able to include every module individually. I'd like to know more about your use case, and why you would want to do that. It was a long time ago, but I remember putting a lot of energy into modularizing the library. Obviously I failed, but I think we can do this now better.

I will have more spare time on weekend, and I'll investigate ffmpeg then (week days are practically work+eat+sleep+repeat). Hopefully I'll also poke around and remember things around DCV and D in general, so I can actually help more. Feels kind of good to be back into this. :)

**Comment by aferust** [comment]
I am happy to cause a skilled programmer to reintroduce the world of d :) Yesterday I made a commit for restructuring source files to keep them in the source folder (except ffmpeg441, see my latest commit). We can do more modularization in the future.

> I'd like to know more about your use case...

Imagine that someone wants to apply a filter or conduct a simple image processing algorithm. One should not need to deal with all ffmpeg-related things for something not related to the video. So, a separate dcv:videoio achieves this.

More about my use case in terms of computer vision:
https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=ferhat+kurtulmu%C5%9F&oq=ferhat+

I have been using image processing and machine learning tools to solve problems in agriculture. I don't have a professional reason for using DCV. I just love D.

I see you are busy. Every one of us is working for paying the bills. Recently, I could find some free time and wanted to be familiar with mir libraries. I think D deserves a CV library at least capable of doing the basic things. I know the community is small and DCV cannot compete with OpenCV any time soon. Maybe DCV attracts more attention from skilled programmers in the future.

**Comment by ljubobratovicrelja** [comment]
>More about my use case in terms of computer vision:
https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=ferhat+kurtulmu%C5%9F&oq=ferhat+

>I have been using image processing and machine learning tools to solve problems in agriculture. I don't have a professional reason for using DCV. I just love D.

This is very interesting! Being in virtual world for the most part in my work, I have huge respect for kind of work you are doing.

Trust me, I understand the passion about D - I've had it as well, and it was my motivation to start DCV. I've stepped away due to my professional life spooling up, and making me leave hobbies like this. Also it was a failed mission to push D through in my company (I was really hopping to marry the two, but I guess D really didn't take off in the industry I am in). Btw I am in computer graphics field, but working a lot on vision systems for scanning shapes and materials (very popular in current graphics / gaming / vfx world), so that was my motivation and aim when originally building DCV.

Anyhow, very happy that you're here Ferhat. Let's give this one another go. :)

Going to inspect your commits right now, and later on looking into ffmpeg-d as promised.

**Comment by aferust** [comment]
Thank you Relja. I am happy that you are here too. Yes, let's do this.

**Comment by ljubobratovicrelja** [comment]
@aferust why split `io` module to its base and `videoio`? I guess because the modularity makes sense, as ffmpeg is such a heavy dependency, therefore people who don't deal with videos, but only images, don't have to worry about it?

Should we for the sake of consistency rename the `io` package to `imageio` then? I think that would logical, and more intuitive for newcomers.

**Comment by ljubobratovicrelja** [comment]
OK, I inherited your changes in [a local branch](https://github.com/libmir/dcv/tree/revision_2022) and integrated ffmpeg-d, so I am closing this PR, and making a new one (sorry for complicating, but I couldn't otherwise make changes myself, afaik at least). Opening up a new PR right after...
